Apache-Beam-Jobs.md

```diff
@@ -227,107 +223,357 @@ error_pcoll = (
+
+```python
+class CountExplorationStatesJobTests(test_jobs.JobTestBase):
+
+    JOB_CLASS = CountExplorationStatesJob
 ```
 
-Finally, we'll import this job into the registry file. Let's assume the name of
-the file was `jobs/count_all_models_jobs.py`.
+### 2. Run assertions using a `assert_job_output_is_*` method
 
-```diff
-  # file: jobs/registry.py
+When testing a job, we should aim to cover behavior and common edge cases. For this job, we'll have 3 main tests:
+1. When there are no Explorations in the datastore.
+2. When there is exactly 1 Exploration in the datastore.
+3. When there are many Explorations in the datastore.
 
-  from jobs import base_jobs
-  from jobs import base_validation_jobs
-+ from jobs import count_all_models_jobs
+```python
+def test_empty_datastore(self):
+    # Don't add any explorations to the datastore.
+    self.assert_job_output_is_empty()
+
+def test_single_exploration(self):
+    self.save_new_linear_exp_with_state_names_and_interactions(
+        'e1', 'o1', ['A', 'B', 'C'], ['TextInput'])
+
+    self.assert_job_output_is([
+        job_run_result.JobRunResult(stdout='3'),
+    ])
+
+def test_many_explorations(self):
+    self.save_new_linear_exp_with_state_names_and_interactions(
+        'e1', 'o1', ['A', 'B', 'C'], ['TextInput'])
+    self.save_new_linear_exp_with_state_names_and_interactions(
+        'e2', 'o1', ['D', 'E', 'F', 'G', 'H'], ['TextInput'])
+    self.save_new_linear_exp_with_state_names_and_interactions(
+        'e3', 'o1', ['I', 'J'], ['TextInput'])
+    self.save_new_linear_exp_with_state_names_and_interactions(
+        'e4', 'o1', ['K', 'L', 'M', 'N'], ['TextInput'])
+
+    self.assert_job_output_is([
+        job_run_result.JobRunResult(stdout='14'),
+    ])
 ```
 
+Note that `self.assert_job_output_is(...)` and `self.assert_job_output_is_empty()` do as advertised -- they run the job to completion and verify the result.
+
+> **IMPORTANT:** Only one `assert_job_output_is` assertion can be performed in a test body. Multiple calls will result in an exception instructing you to split the test apart.
+
+Just because a job passes in unit tests does not guarantee it will pass in production. This is because workers, which execute the pipeline code, are run in a special environment where the code base is configured differently. While Oppia's jobs team works to resolve the differences, be careful about using complex and/or confusing objects. The simpler your job, the greater chance it'll work in production!
+
+## Running Apache Beam Jobs
+
+### Local Development Server
+
+These instructions assume you are running a local development server. If you are a release coordinator running these jobs on the production or testing servers, you should already have been granted the "Release Coordinator" role, so you can skip steps 1-3.
```

Comment:
Should we mention somewhere that Beam validates Python types and that they shouldn't be left as the last step?

Reply:
Done.